lapce-data/src/hover.rs

```diff
@@ -152,41 +170,317 @@ impl Default for HoverData {
+
+#[derive(Data, Clone)]
+pub struct MarkdownText {
+    /// Rendered rich text
+    text: RichText,
+    /// Indices into the underlying text that code blocks start and begin at
+    /// This lets us calculate a rectangle surrounding the text to apply a background too, in order
+    /// to differentiate it from normal text more.
+    /// Note that this covers multiline and inline codeblocks
+    #[data(eq)]
+    code_block_indices: Vec<Range<usize>>,
+}
+impl MarkdownText {
+    fn new(text: RichText, code_block_indices: Vec<Range<usize>>) -> Self {
+        Self {
+            text,
+            code_block_indices,
+        }
+    }
+
+    pub fn empty() -> Self {
+        Self {
+            text: RichText::new(ArcStr::from("")),
+            code_block_indices: Vec::new(),
+        }
+    }
+
+    /// Draw the markdown to the paint ctx, using the layout
+    pub fn draw(
+        &self,
+        ctx: &mut PaintCtx,
+        env: &Env,
+        origin: Point,
+        layout: &TextLayout<MarkdownText>,
+        code_background_color: Color,
+    ) {
+        let rect = ctx.region().bounding_box();
+        for range in &self.code_block_indices {
+            // TODO: We could make a function to get just p0, since we don't need the other
+            // calculated point
+            let start_point = layout.cursor_line_for_text_position(range.start).p0;
+            let end_point = layout.cursor_line_for_text_position(range.end).p0;
+
+            // If the end_point and start_point x positions are the same (typically 0), then
+            // they're probably a multiline code block.
+            let end_x = if end_point.x == start_point.x {
+                // Use the entire width, avoiding the scrollbar to make it appear nicer
+                rect.x1
+                    - env.get(theme::SCROLLBAR_WIDTH)
+                    - env.get(theme::SCROLLBAR_PAD)
+            } else {
+                origin.x + end_point.x
+            };
+
+            let bg_rect = Rect::new(
+                origin.x + start_point.x,
+                origin.y + start_point.y,
+                end_x,
+                origin.y + end_point.y,
+            );
+            ctx.fill(bg_rect, &code_background_color);
+        }
+        layout.draw(ctx, origin);
+    }
+}
+// Let this be treated as Text
+impl PietTextStorage for MarkdownText {
+    fn as_str(&self) -> &str {
+        self.text.as_str()
+    }
+}
+impl TextStorage for MarkdownText {
+    fn add_attributes(
+        &self,
+        builder: druid::piet::PietTextLayoutBuilder,
+        env: &druid::Env,
+    ) -> druid::piet::PietTextLayoutBuilder {
+        self.text.add_attributes(builder, env)
+    }
+
+    fn env_update(&self, ctx: &druid::text::EnvUpdateCtx) -> bool {
+        self.text.env_update(ctx)
+    }
+
+    fn links(&self) -> &[druid::text::Link] {
+        self.text.links()
+    }
+}
+
+// TODO: This could moved into a more general utility file, for use in things other than hover
```

Comment:
Maybe most of the markdown-specific code could be moved out of hover. But looking at how flat Lapce's source code is in terms of modules, this might be done at a later stage. However, I think some todos could be added to direct some attention to this.

Reply:
Changed the comment a bit and moved it before most of the markdown parsing/rendering code. Though I'm unsure what you mean with adding more todos.

Reply:
> I'm unsure what you mean

Basically what you did. I meant to mark all the relevant structs, but now it's plenty good, thanks.